src/mlpack/methods/ann/layer/reparametrization_impl.hpp

```diff
@@ -0,0 +1,117 @@
+/**
+ * @file reparametrization_impl.hpp
+ * @author Atharva Khandait
+ *
+ * Implementation of the Reparametrization layer class which samples from a
+ * gaussian distribution.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_METHODS_ANN_LAYER_REPARAMETRIZATION_IMPL_HPP
+#define MLPACK_METHODS_ANN_LAYER_REPARAMETRIZATION_IMPL_HPP
+
+// In case it hasn't yet been included.
+#include "reparametrization.hpp"
+
+namespace mlpack {
+namespace ann /** Artificial Neural Network. */ {
+
+template<typename InputDataType, typename OutputDataType>
+Reparametrization<InputDataType, OutputDataType>::Reparametrization() :
+    latentSize(0),
+    stochastic(true),
+    includeKl(true)
+{
+  // Nothing to do here.
+}
+
+template <typename InputDataType, typename OutputDataType>
+Reparametrization<InputDataType, OutputDataType>::Reparametrization(
+    const size_t latentSize,
+    const bool stochastic,
+    const bool includeKl) :
+    latentSize(latentSize),
+    stochastic(stochastic),
+    includeKl(includeKl)
+{
+  // Nothing to do here.
+}
+
+template<typename InputDataType, typename OutputDataType>
+template<typename eT>
+void Reparametrization<InputDataType, OutputDataType>::Forward(
+    const arma::Mat<eT>&& input, arma::Mat<eT>&& output)
+{
+  if (input.n_rows != 2 * latentSize)
+  {
+    Log::Fatal << "The output size of layer before the Reparametrization "
+        << "layer should be 2 * latent size of the Reparametrization layer!"
+        << std::endl;
+  }
+
+  mean = input.submat(latentSize, 0, 2 * latentSize - 1, input.n_cols - 1);
+  preStdDev = input.submat(0, 0, latentSize - 1, input.n_cols - 1);
+
+  if (stochastic)
+    gaussianSample = arma::randn<arma::Mat<eT> >(latentSize, input.n_cols);
+  else
+    gaussianSample = arma::ones<arma::Mat<eT> >(latentSize, input.n_cols) * 0.7;
+
+  SoftplusFunction::Fn(preStdDev, stdDev);
+  output = mean + stdDev % gaussianSample;
+}
+
+template<typename InputDataType, typename OutputDataType>
+template<typename eT>
+void Reparametrization<InputDataType, OutputDataType>::Backward(
+    const arma::Mat<eT>&& /* input */, arma::Mat<eT>&& gy, arma::Mat<eT>&& g)
+{
+  SoftplusFunction::Deriv(preStdDev, g);
+
+  if (includeKl)
+  {
+    arma::Mat<eT> klBack;
+    klBackward(std::move(klBack));
+    g = join_cols(gy % std::move(gaussianSample) % g, gy) + std::move(klBack);
+  }
+  else
+    g = join_cols(gy % std::move(gaussianSample) % g, gy);
+}
+
+template<typename InputDataType, typename OutputDataType>
+template<typename InputType>
+double Reparametrization<InputDataType, OutputDataType>::klForward(
+    const InputType&& input)
```

Comment:
Can we rename this KLForward() to match the style guide?  Alternately, it is such a small function that it could be inlined into `Loss()` since it is only used there.

The same applies for `klBackward()` too.

Reply:
I will make the style changes, sorry for not keeping them in mind. Although, I wanted `klForward()` and `klBackward()` to be seperate from the `Forward()` and `Backward()` of the main Repar layer. For some other reasons I implemented them here. Is that okay? Will not having this inlined into Loss() cause significant performance issues? If yes then I will make the changes.

Reply:
Depending on the size I think it could be useful to inline the function and to avoid ``join_cols`` and the corresponding split afterwards.